src/rust/engine/async_semaphore/src/lib.rs

```diff
@@ -97,30 +121,64 @@ impl Drop for Permit {
   }
 }
 
+#[derive(Clone)]
 pub struct PermitFuture {
   inner: Option<Arc<Mutex<Inner>>>,
+  task_id: Option<usize>,
+}
+
+impl Drop for PermitFuture {
+  fn drop(&mut self) {
+    // if task_id is Some then this PermitFuture was added to the waiters queue.
+    if let Some(task_id) = self.task_id {
+      let inner = self.inner.take().unwrap();
+      let mut inner = inner.lock();
+      if let Some(task_index) = inner.waiters.iter().position(|task| task_id == task.id) {
+        inner.waiters.remove(task_index);
+      }
+    }
+  }
 }
 
 impl Future for PermitFuture {
   type Item = Permit;
   type Error = ();
 
   fn poll(&mut self) -> Poll<Permit, ()> {
-    let inner = self.inner.take().expect("cannot poll PermitFuture twice");
+    let inner = self.inner.clone().unwrap();
     let acquired = {
       let mut inner = inner.lock();
       if inner.available_permits == 0 {
-        inner.waiters.push_back(task::current());
+        if self.task_id.is_none() {
+          let task_id = inner.task_id;
+          let this_task = UniqueTask {
+            id: inner.task_id,
+            task: task::current(),
+          };
+          self.task_id = Some(task_id);
+          inner.task_id += 1;
+          inner.waiters.push_back(this_task);
+          trace!(
+            "Added a task to the queue number of waiters is {:?}.",
+            inner.waiters.len()
+          );
+        }
         false
       } else {
+        if let Some(front_task) = inner.waiters.front() {
+          // This task is the one we notified, so remove it. Otherwise keep it on the
+          // waiters queue so that it doesn't get forgotten.
+          if front_task.task.will_notify_current() {
+            inner.waiters.pop_front();
+          }
+        }
         inner.available_permits -= 1;
```

Comment:
Should this be inside of the `will_notify_current` block...? Only the "correct" task (the one that was notified) should actually acquire and decrement the permits?

Reply:
I think that makes sense. I will change it and see if build still pass :)

Reply:
from the tokio docs
>As this is purely a performance optimization a valid implementation for this function is to always return false. A best effort is done to return true where possible, but false negatives may happen. Note that this function will not return a false positive, however.

Unfortunately I think this means we have to let whoever is polling get access to the semaphore if there is a permit available. In the pathological case where `will_notify_current` always returns false we would stop issuing permits to pollers.

Reply:
Making this change causes the async_semaphore test suite to be flaky which to me is an indication that the `will_notify_current` function is also in practice a "best effort".

Reply:
This issue of task identity not being stable is probably they main reason we should talk to the tokio folks about a more "correct" implementation of the semaphore for futures.

Reply:
@stuhood just want to make sure you see this.

Reply:
But... that means we're not actually using the task_id to determine whether we were supposed to be the task that was removed from the semaphore.

I think that my understanding of how we were trying to achieve fairness here was for the task to see whether it was the head of the queue: we're currently peeking at the queue, and then acquiring the semaphore regardless of whether we ("the task") were at the head of the queue. It's fine that `will_notify_current` is best effort, because shouldn't we be comparing the task id of the head of the queue to our own task id?

Also, I think that part of the confusion here might be that "task id" here is not the id of the task (can't be). Instead, it's the id of a particular PermitFuture. So... maybe this should be acquire_attempt_id or something?

Reply:
That is a great point about comparing the id of the PermitFuture to the UniqueTask in the queue, I'm not sure why I didn't see that. Anyway, I will implement that and see what happens.